Refactor next version calculator (get rid of `taggedSemanticVersion`)

AlexPykavy opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

Suggest to review by commits :) I tried to make them as atomic as possible.

## Description
<!--- Describe your changes in detail -->
Minor changes:
- Rearranging code 
- Renaming variables
- Reusing existing functionality

Change next version calculator to act according to the next algorithm:
1. Calculate next version candidate based on the base one
2. Look for the latest `PreReleaseTag` matching Major, Minor and Patch of the next version candidate.
3. Append the `PreReleaseTag`, if found, to the next version candidate.
4. Compare the next version candidate with `Context.CurrentCommitTaggedVersion` including the `PreReleaseTag` comparison.
5. Increment the `PreReleaseTag` of the next version candidate if versions mismatch.
6. Return the next version candidate.

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

I found `NextVersionCalculator` logic complicated and decided to simplify it.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

By running all existing tests.

## Screenshots (if appropriate):

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

HHobeck commented:
I like your refactoring Alex. 👍 Good work. When I execute the unit tests I got some NullReferenceExceptions.

AlexPykavy commented:
@HHobeck , all failing tests are due to the .NET 7 instability on ARM. Can this PR be merged or it needs some corrections?

arturcic commented:
@AlexPykavy - can you please fix the warnings as well? https://github.com/GitTools/GitVersion/actions/runs/3577000238

AlexPykavy commented:
> @AlexPykavy - can you please fix the warnings as well? https://github.com/GitTools/GitVersion/actions/runs/3577000238

Fixed, but I'm not sure about adding `Context.CurrentCommitTaggedVersion != null` before calling `semver.CompareTo(Context.CurrentCommitTaggedVersion)` because the method [handles `null` arguments](https://github.com/GitTools/GitVersion/blob/9895652756df7cb19cc46b6646f802f2fdffeb4a/src/GitVersion.Core/VersionCalculation/SemanticVersioning/SemanticVersion.cs#L215-L218). Maybe it would be better to use null-forgiving operator to suppress the warning?

HHobeck commented:
> > @AlexPykavy - can you please fix the warnings as well? https://github.com/GitTools/GitVersion/actions/runs/3577000238
> 
> Fixed, but I'm not sure about adding `Context.CurrentCommitTaggedVersion != null` before calling `semver.CompareTo(Context.CurrentCommitTaggedVersion)` because the method [handles `null` arguments](https://github.com/GitTools/GitVersion/blob/9895652756df7cb19cc46b6646f802f2fdffeb4a/src/GitVersion.Core/VersionCalculation/SemanticVersioning/SemanticVersion.cs#L215-L218). Maybe it would be better to use null-forgiving operator to suppress the warning?

This might be also possible:
```csharp
if (Context.CurrentCommitTaggedVersion?.CompareTo(semver) == 0)
{
...
}
```

HHobeck commented on src/GitVersion.Core/VersionCalculation/NextVersionCalculator.cs:
```diff
@@ -47,88 +47,70 @@ public virtual NextVersion FindVersion()
-        var branchConfigHasPreReleaseTagConfigured = !tag.IsNullOrEmpty();
-        var preReleaseTagDoesNotMatchConfiguration = hasPreReleaseTag && branchConfigHasPreReleaseTagConfigured && semver.PreReleaseTag?.Name != tag;
-        if (semver.PreReleaseTag?.HasTag() != true && branchConfigHasPreReleaseTagConfigured || preReleaseTagDoesNotMatchConfiguration)
+        if (Context.CurrentCommitTaggedVersion != null && semver.CompareTo(Context.CurrentCommitTaggedVersion) == 0)
```
`Context.CurrentCommitTaggedVersion?.CompareTo(semver) == 0` would be okay. It is also fine with me if you set the parameter data type in CompareTo and allowing `null` values.

AlexPykavy replied:
Corrected `CompateTo(SemanticVersion)` method to accept nullable argument because its overload (`CompateTo(SemanticVersion?, bool)` already handles it.

HHobeck commented on src/GitVersion.Core/VersionCalculation/NextVersionCalculator.cs:
```diff
@@ -47,88 +47,70 @@ public virtual NextVersion FindVersion()
+        if (Context.CurrentCommitTaggedVersion != null && semver.CompareTo(Context.CurrentCommitTaggedVersion) == 0)
         {
-            UpdatePreReleaseTag(new(nextVersion.Branch, nextVersion.Configuration), semver, nextVersion.BaseVersion.BranchNameOverride);
+            if (semver.BuildMetaData != null)
```
This is always true because you have set it before with a not null value (see comment before). Did we lose a use case!? It's fine with me if you delete this check as far the unit tests are green. ;)

HHobeck replied:
I would say: Change the data type of the property PreReleaseTag and BuildMetaData to a non-nullable type in the class SemanticVersion. ;)

AlexPykavy replied:
Done 🤠

HHobeck commented on src/GitVersion.Core/VersionCalculation/NextVersionCalculator.cs:
```diff
@@ -47,88 +47,70 @@ public virtual NextVersion FindVersion()
             EnsureHeadIsNotDetached(Context);
         }
 
-        // TODO: It is totally unimportant that the current commit has been tagged or not IMO. We can make a double check actually if the result
```
Very nice refactoring. Thank you!

HHobeck commented:
> @HHobeck , all failing tests are due to the .NET 7 instability on ARM. Can this PR be merged or it needs some corrections?

I have the same issues on the other pull-request branches. Might be an idea to disable the tests temporally. @arturcic what do you think?

HHobeck commented:
@arturcic: May I ask you is it possible to go away from .net standard 2.1? It makes no sens in my opinion if we move to .net 6 and 7 or I'm wrong? It would be nice to use the Init Only Setters.

arturcic commented:
Honestly I was considering that. We need to check the libgit2sharp if can pe used in this case

HHobeck commented:
> libgit2sharp

But libgit2sharp is only referenced in GitVersion.App.csproj and not in GitVersion.Core.csproj

HHobeck commented:
@AlexPykavy Could you please take a look on all references of the property `PreReleaseTag` and `BuildMetaData` and remove the following null checks?
- e.g. PreReleaseTag?.HasTag(), PreReleaseTag?.Number, PreReleaseTag == null and PreReleaseTag.ShouldNotBeNull()
- e.g. BuildMetaData?.Sha, BuildMetaData == null and PreReleaseTag.ShouldNotBeNull()

HHobeck commented:
From my point of view the next refactoring steps are the following:

- removing the .net standard target framework and change it to .net 6.0 and .net 7.0
- fix all warnings because of the nullable feature (which is enabled by default)
- change the SemanticVersionPreReleaseTag and SemanticVersionBuildMetaData to a struct data type and make it immutable
- change the properties in SemanticVersionPreReleaseTag and SemanticVersionBuildMetaData to read only using Init Only Setters feature of c#
- remove all mock setups which returning empty SemanticVersionPreReleaseTag and SemanticVersionBuildMetaData

What do you think?

arturcic commented:
@HHobeck I can take this one in a separate PR:
> removing the .net standard target framework and change it to .net 6.0 and .net 7.0

As for the others, you can work on them if you have a bit of time

AlexPykavy commented:
> @AlexPykavy Could you please take a look on all references of the property `PreReleaseTag` and `BuildMetaData` and remove the following null checks?
> 
> * e.g. PreReleaseTag?.HasTag(), PreReleaseTag?.Number, PreReleaseTag == null and PreReleaseTag.ShouldNotBeNull()
> * e.g. BuildMetaData?.Sha, BuildMetaData == null and PreReleaseTag.ShouldNotBeNull()

Done if I understood you correctly :)

AlexPykavy commented:
@HHobeck, updated. You got me there: I forgot about checking `!= null`.

HHobeck commented:
Okay thank your ver much @AlexPykavy for your work!! :) From my point of view the PR can be merged to main. @arturcic @asbjornu

HHobeck commented:
I'm not sure why the second tests fails. Maybe you can help me? @arturcic, @AlexPykavy

Is it is because of the refactoring in the PR? This two tests are new and comming from the main branch:

![image](https://user-images.githubusercontent.com/56404113/219942542-6ae003f4-162b-40e6-8c58-95258c38b4a6.png)

Message: 
Shouldly.ShouldAssertException : variables.FullSemVer
    should be
"1.4.0"
    but was
"1.4.0-rc"

AlexPykavy commented:
Fixed by considering empty `preRelaseTag` as a edge case and moving out the corresponding `if` block. I've also rebased my previous PR on top of current `main` instead of merging because I like prefer rebase in such cases :)

arturcic commented:
@HHobeck do you approve the changes? In case yes we can merge it

arturcic commented:
@HHobeck do you approve the changes? In case yes we can merge it

arturcic commented:
@HHobeck do you approve the changes? In case yes we can merge it

arturcic commented:
@HHobeck do you approve the changes? In case yes we can merge it

arturcic commented:
@HHobeck do you approve the changes? In case yes we can merge it

arturcic commented:
@HHobeck do you approve the changes? In case yes we can merge it

arturcic commented:
@HHobeck do you approve the changes? In case yes we can merge it

arturcic commented:
<!-- GitReleaseManager release comment -->
:tada: This issue has been resolved in version 6.0.0-beta.1 :tada:
The release is available on:
- [GitHub Release](https://github.com/GitTools/GitVersion/releases/tag/6.0.0-beta.1)
- [Chocolatey Package](https://chocolatey.org/packages/GitVersion.Portable/6.0.0-beta.1)
- [.Net Global Tool](https://www.nuget.org/packages/GitVersion.Tool/6.0.0-beta.1)

Your **[GitReleaseManager](https://github.com/GitTools/GitReleaseManager)** bot :package::rocket:
